### PR TITLE
Activity submit edit and update in database.

### LINKED ISSUE #48: Save data from "edit activities" form
### Issue summary

Need to save the data that the user inputs into the "edit activity" form.

### Expected behavior

Data should be validated and updated in the database.

### Actual behavior

Data is discarded.

### PR DESCRIPTION
**Pull request by ananyeet**
### What is a quick description of the change?

Make sure the database updates with new entries via user input.

### Is this fixing an issue?

fixes #48 

### Are there more details that are relevant?

Created #64.

### Check lists (check `x` in `[ ]` of list items)

- [  ] Test written/updating
- [  ] Tests passing
- [  ] Coding style (indentation, etc)

Please explain why any are not present, if any.

### Any additional comments?

{Fill in here or remove (leave question)}

### CODE REVIEW

File: frontend/src/components/ViewActivities/activityfns.js [hunk 1]
```diff
@@ -54,4 +57,30 @@ export function getField(activity, fieldName, defaultValue){
...
+/**
+ * Write contents into an activity in the database.
+ * 
+ * @param {string} tripId Database ID of the trip whose actiivty should be modified.
+ * @param {string} activityId Database ID of the activity to be modified.
+ * @param {Object} newValues Dictionary of the new values in {fieldName: newValue} form
+ */
+export async function writeActivity(tripId, activityId, newValues) {
+  // todo: check if tripId or activityId is not valid. (#58)
+  if (newValues === null) {
```
Comment (zghera) [inline frontend/src/components/ViewActivities/activityfns.js:71]: Is this ever possible? I pretty sure that at a minimum it will just be an object with at least one key:val pair given your instantiation of `newVals` and the call to `writeActivity` in `activity.js`.

File: frontend/src/components/ViewActivities/activityfns.js [hunk 2]
```diff
@@ -54,4 +57,30 @@ export function getField(activity, fieldName, defaultValue){
...
   return activity[fieldName];
+}
+
+/**
+ * Write contents into an activity in the database.
+ * 
+ * @param {string} tripId Database ID of the trip whose actiivty should be modified.
+ * @param {string} activityId Database ID of the activity to be modified.
+ * @param {Object} newValues Dictionary of the new values in {fieldName: newValue} form
+ */
```
Comment (zghera) [inline frontend/src/components/ViewActivities/activityfns.js:68]: include the @return {boolean} ...

File: frontend/src/components/ViewActivities/activityfns.js [hunk 3]
```diff
@@ -54,4 +57,30 @@ export function getField(activity, fieldName, defaultValue){
...
+  const act = db.collection(DB.COLLECTION_TRIPS).doc(tripId)
+    .collection(DB.COLLECTION_ACTIVITIES).doc(activityId);
+  
+  try {
+    // Note: newValues cannot contain lists. Check documentation for update().
+    const a = await act.update(newValues);
+    return a ? true : false;
+  } catch (e) {
+    console.log(e);
+    return false;
```
Comment (zghera) [inline frontend/src/components/ViewActivities/activityfns.js:84]: Just curious, why return a promise holding a boolean rather than just returning the result of the update?

File: frontend/src/components/ViewActivities/activity.js [hunk 4]
```diff
@@ -40,7 +50,28 @@ class Activity extends React.Component {
...
-  finishEditActivity = () => (  this.setState({editing: false}) );
+  finishEditActivity(event) {
+    this.setState({editing: false});
+    event.preventDefault();
+    this.editActivity();
+  };
+
+  /**
+   * Edit an activity in the database upon form submission.
+   * TODO: Update times as well! This only does the text field forms.
```
[the remark this answers is not in the document]
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activity.js:61]: yup #64

File: frontend/src/components/ViewActivities/activity.js [hunk 5]
```diff
@@ -40,7 +50,28 @@ class Activity extends React.Component {
...
+   */
+  editActivity() {
+    let newVals = {};
+    if (this.editTitleRef.current !== null) {
```
[the remark this answers is not in the document]
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activity.js:65]: Yeah, it's null if you leave the field blank.

File: frontend/src/components/ViewActivities/activityfns.js (hunk 1 above)
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js:71]: Yeah, if the user just looks at a trip and then submits the form without changing anything. I'll put a "timestamp" field in there just so it's never empty.

File: frontend/src/components/ViewActivities/activityfns.js (hunk 3 above)
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js:84]: That does return the boolean. `a ? true: false` returns `true` if `a` exists and `false` otherwise.

File: frontend/src/components/ViewActivities/activityfns.js (hunk 1 above)
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js:71]: nope. If the user submits an empty form .current.value is "". Fixed.

File: frontend/src/components/ViewActivities/activity.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activity.js]: nope. fixed.

File: frontend/src/components/ViewActivities/activityfns.js [hunk 6]
```diff
@@ -54,4 +58,32 @@ export function getField(activity, fieldName, defaultValue){
...
+ * @returns true if the write was successful, false otherwise.
+ */
+export async function writeActivity(tripId, activityId, newValues) {
+  // todo: check if tripId or activityId is not valid. (#58)
+  newValues = {
+    ...newValues, 
+    'fillerstamp': Firebase.firestore.Timestamp.now()
+  };
+
+  const act = db.collection(DB.COLLECTION_TRIPS).doc(tripId)
```
Comment (keiffer01) [inline frontend/src/components/ViewActivities/activityfns.js:78]: Nit: call this 'activity' instead?
```suggestion
  const activity = db.collection(DB.COLLECTION_TRIPS).doc(tripId)
```

File: frontend/src/components/ViewActivities/activityfns.js [hunk 7]
```diff
@@ -54,4 +58,32 @@ export function getField(activity, fieldName, defaultValue){
...
+    .collection(DB.COLLECTION_ACTIVITIES).doc(activityId);
+  
+  try {
+    // Note: newValues cannot contain lists. Check documentation for update().
+    const a = await act.update(newValues);
+    return a ? true : false;
```
Comment (keiffer01) [inline frontend/src/components/ViewActivities/activityfns.js:84]: I might be missing something but what is the await and returning of a boolean used for?

Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js:84]: return a boolean to tell whether or not the write was successful.

Comment (zghera) [reply frontend/src/components/ViewActivities/activityfns.js:84]: Did you see what `a` was in either case (did or did not exist) because I think it may be undefined in either case given that [update](https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentReference#update) returns a void promise (I could be wrong but I found [this](https://stackoverflow.com/questions/44216949/does-a-async-function-which-returns-promisevoid-have-an-implicit-return-at-the) when trying to confirm my suspicions).

File: frontend/src/components/ViewActivities/activityfns.js [hunk 8]
```diff
@@ -54,4 +58,32 @@ export function getField(activity, fieldName, defaultValue){
...
+    ...newValues, 
+    'fillerstamp': Firebase.firestore.Timestamp.now()
+  };
+
+  const activity = db.collection(DB.COLLECTION_TRIPS).doc(tripId)
+    .collection(DB.COLLECTION_ACTIVITIES).doc(activityId);
+  
+  try {
+    // Note: newValues cannot contain lists. Check documentation for update().
+    const a = await act.update(newValues);
```
Comment (zghera) [inline frontend/src/components/ViewActivities/activityfns.js:83]: Did you see what `a` was in either case (did or did not exist) because I think it may be undefined in either case given that [update](https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentReference#update) returns a void promise (I could be wrong but I found [this](https://stackoverflow.com/questions/44216949/does-a-async-function-which-returns-promisevoid-have-an-implicit-return-at-the) when trying to confirm my suspicions).

Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js]: I found something contradictory, but I can't find it anymore. Fixed.

File: frontend/src/components/ViewActivities/activity.js [hunk 9]
```diff
@@ -57,30 +88,31 @@ class Activity extends React.Component {
...
+          <Form.Group as={Row} controlId='formActivityStartTime'>
             <Col sm={2}><Form.Label>From:</Form.Label></Col>
-            <Col sm={4}><Form.Control type="date" label="date"/></Col>
-            <Col sm={2}><Form.Control type="time" label="time"/></Col>
+            <Col sm={4}><Form.Control type='date' label='date' ref={this.editStartDateRef}/></Col>
+            <Col sm={2}><Form.Control type='time' label='time' ref={this.editStartTimeRef}/></Col>
             <Col sm={1}>{timezonePicker()}</Col>
           </Form.Group>
-          <Form.Group as={Row} controlId="formActivityStartTime">
+          <Form.Group as={Row} controlId='formActivityStartTime'>
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activity.js:102]: Should this be 'formActivityEndTime'?

File: frontend/src/components/ViewActivities/activity.js [hunk 10]
```diff
@@ -57,30 +88,31 @@ class Activity extends React.Component {
...
             <Col sm={2}><Form.Label>To:</Form.Label></Col>
-            <Col sm={4}><Form.Control type="date" label="date"/></Col>
-            <Col sm={2}><Form.Control type="time" label="time"/></Col>
+            <Col sm={4}><Form.Control type='date' label='date' ref={this.editEndDateRef}/></Col>
+            <Col sm={2}><Form.Control type='time' label='time' ref={this.editEndTimeRef}/></Col>
             <Col sm={1}>{timezonePicker()}</Col>
           </Form.Group>
-          <Form.Group as={Row} controlId="formActivityTitle">
+          <Form.Group as={Row} controlId='formActivityTitle'>
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activity.js:108]: Should this be 'formActivityDescription'?

File: frontend/src/components/ViewActivities/activityfns.js (hunk 2 above)
Comment (Hiram-Silvey) [reply frontend/src/components/ViewActivities/activityfns.js:70]: +1, this should include the return type "{boolean}" rather than just "true."

File: frontend/src/components/ViewActivities/activityfns.js [hunk 11]
```diff
@@ -54,4 +59,32 @@ export function getField(activity, fieldName, defaultValue){
...
\ No newline at end of file
+}
+
+/**
+ * Write contents into an activity in the database.
+ * 
+ * @param {string} tripId Database ID of the trip whose actiivty should be modified.
+ * @param {string} activityId Database ID of the activity to be modified.
+ * @param {Object} newValues Dictionary of the new values in {fieldName: newValue} form
+ * @returns true if the write was successful, false otherwise.
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activityfns.js:70]: Add type.

File: frontend/src/components/ViewActivities/activityfns.js [hunk 12]
```diff
@@ -44,14 +69,42 @@ export function compareActivities(a, b) {
...
   return activity[fieldName];
+}
+
+/**
+ * Write contents into an activity already existing in the database.
+ * 
+ * @param {string} tripId Database ID of the trip whose actiivty should be modified.
+ * @param {string} activityId Database ID of the activity to be modified.
+ * @param {Object} newValues Dictionary of the new values in <code>{fieldName: newValue}</code> form
+ * @returns {boolean} <code>true</code> if the write was successful, <code>false</code> otherwise.
```
Comment (beiebi0) [inline frontend/src/components/ViewActivities/activityfns.js:90]: Is this return value being used anywhere?

Comment (ananyeet) [reply frontend/src/components/ViewActivities/activityfns.js:90]: not officially, no. could be used for error messages later though

File: frontend/src/components/ViewActivities/activity.js [hunk 13]
```diff
@@ -107,7 +111,7 @@ class Activity extends React.Component {
             <Col sm={2}><Form.Control type='time' label='time' ref={this.editEndTimeRef}/></Col>
             <Col sm={1}>{timezonePicker()}</Col>
           </Form.Group>
-          <Form.Group as={Row} controlId='formActivityDescription'>
+          <Form.Group as={Row} controlId="formActivityDescription">
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activity.js:114]: Why is this changed to double quotes?

File: frontend/src/components/ViewActivities/activityday.js [hunk 14]
```diff
@@ -20,7 +20,7 @@ class ActivityDay extends React.Component {
       <div className='activity-day'>
         <h4>{time.timestampToDateFormatted(date.getTime())}</h4>
         {sortedActivities.map((activity, index) => (
-          <Activity activity={activity} key={index} className='activity'/>
+          <Activity activity={activity} key={index} className="activity"/>
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activityday.js]: Same as above.

File: frontend/src/components/ViewActivities/activityfns.js [hunk 15]
```diff
@@ -62,12 +82,12 @@ export function getField(activity, fieldName, defaultValue) {
...
 /**
- * Write contents into an activity in the database.
+ * Write contents into an activity already existing in the database.
  * 
  * @param {string} tripId Database ID of the trip whose actiivty should be modified.
  * @param {string} activityId Database ID of the activity to be modified.
- * @param {Object} newValues Dictionary of the new values in {fieldName: newValue} form
- * @returns true if the write was successful, false otherwise. 
+ * @param {Object} newValues Dictionary of the new values in <code>{fieldName: newValue}</code> form
+ * @returns {boolean} <code>true</code> if the write was successful, <code>false</code> otherwise.
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activityfns.js:90]: Update to @return.

File: frontend/src/components/ViewActivities/activityfns.js [hunk 16]
```diff
@@ -49,10 +69,10 @@ export function compareActivities(a, b) {
...
  * Get the field of field name `fieldName` from `activity` or the default value.
  * 
- * @param {Object} activity 
- * @param {string} fieldName 
- * @param defaultValue 
- * @returns `activity[fieldName]` if possible, else `defaultValue`.
+ * @param {ActivityInfo} activity The activity from which to get the field.
+ * @param {string} fieldName Name of field to get.
+ * @param {*} defaultValue Value if field is not found/is null.
+ * @returns {*} <code>activity[fieldName]</code> if possible, else <code>defaultValue</code>.
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewActivities/activityfns.js:75]: Update to @return.

File: frontend/src/components/ViewActivities/activity.js (hunk 13 above)
Comment (ananyeet) [reply frontend/src/components/ViewActivities/activity.js:114]: probably from a merge that I did :/
